## Final Project Submission

Please fill out:
* Student name: 
* Student pace: self paced / part time / full time
* Scheduled project review date/time: 
* Instructor name: 
* Blog post URL:


In [1]:
# Your code here - remember to use markdown cells for comments as well!

In [13]:
# Import the relevant libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
import scipy.stats as stats
import seaborn as sns
%matplotlib inline


In [3]:
# Load the data
df = pd.read_csv(r'C:\Users\Arnold Chovu\Desktop\Data_Science\Projects\dsc-phase-2-project-v2-3\data\kc_house_data.csv')
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors waterfront  view  ...          grade sqft_above  \
0      5650     1.0        NaN  NONE  ...      7 Average       1180   
1      7242     2.0         NO  NONE  ...      7 Average       2170   
2     10000     1.0         NO  NONE  ...  6 Low Average        770   
3      5000     1.0         NO  NONE  ...      7 Average       1050   
4      8080     1.0         NO  NONE  ...         8 Good       1680   

   sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0            0.0     1955           0.0    98178  47.5112 -122.257   
1          400.0     1951        1991.0    98125  47.7210 -122.319   
2            0.0     1933           NaN    98028  47.7379 -122.233   
3          910.0     1965           0.0    98136  47.5208 -122.393   
4            0.0     1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 21 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  object 
 9   view           21534 non-null  object 
 10  condition      21597 non-null  object 
 11  grade          21597 non-null  object 
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

In [5]:
df.shape

(21597, 21)

# Data Understanding

The data used in this analysis is for a King County Housing price. The data essentially contains the data that can be used by any stakeholder who is looking to buy a house in King County. The data set contains 21597 entries and 21 rows. The rows contains elements that describe the house in depth, i.e., number of bedrooms, the area of the house, zipcode, year built, the are of the basement and many more characteristics. This features described by the data set can be really helpful for individuals looking to buy houses in King County. The data can show the relationship of the features in the dataset and the price of the houses. However, before the data can be used, we have to perfomr data cleaning to ensure we are working with the right data hence coming up with relevant conclusions. 

## Data Cleaning

#### Transforming the data types

The first observation I made on the data types is the fact that sqft_basement is an object rather than a float or an integer. The measurement of are is always in numerical form. I will convert the data to a float so that it can also be used in analysis. 

In [7]:
# Check the unique values in sqft_basement to find the '?'
df.sqft_basement.unique()

array(['0.0', '400.0', '910.0', '1530.0', '?', '730.0', '1700.0', '300.0',
       '970.0', '760.0', '720.0', '700.0', '820.0', '780.0', '790.0',
       '330.0', '1620.0', '360.0', '588.0', '1510.0', '410.0', '990.0',
       '600.0', '560.0', '550.0', '1000.0', '1600.0', '500.0', '1040.0',
       '880.0', '1010.0', '240.0', '265.0', '290.0', '800.0', '540.0',
       '710.0', '840.0', '380.0', '770.0', '480.0', '570.0', '1490.0',
       '620.0', '1250.0', '1270.0', '120.0', '650.0', '180.0', '1130.0',
       '450.0', '1640.0', '1460.0', '1020.0', '1030.0', '750.0', '640.0',
       '1070.0', '490.0', '1310.0', '630.0', '2000.0', '390.0', '430.0',
       '850.0', '210.0', '1430.0', '1950.0', '440.0', '220.0', '1160.0',
       '860.0', '580.0', '2060.0', '1820.0', '1180.0', '200.0', '1150.0',
       '1200.0', '680.0', '530.0', '1450.0', '1170.0', '1080.0', '960.0',
       '280.0', '870.0', '1100.0', '460.0', '1400.0', '660.0', '1220.0',
       '900.0', '420.0', '1580.0', '1380.0', '475.0', 

We can see the '?' hidden between the values. This might be the character that led to the columne being encoded as an object rather than an integer or a float.

In [8]:
# Drop the row with the question mark since it is a negligible number of rows
df = df[df['sqft_basement'] != '?']

In [9]:
# Check if the '?' is dropped
df.sqft_basement.unique()

array(['0.0', '400.0', '910.0', '1530.0', '730.0', '1700.0', '300.0',
       '970.0', '760.0', '720.0', '700.0', '820.0', '780.0', '790.0',
       '330.0', '1620.0', '360.0', '588.0', '1510.0', '410.0', '990.0',
       '600.0', '560.0', '550.0', '1000.0', '1600.0', '500.0', '1040.0',
       '880.0', '1010.0', '240.0', '265.0', '290.0', '800.0', '540.0',
       '710.0', '840.0', '380.0', '770.0', '480.0', '570.0', '1490.0',
       '620.0', '1250.0', '1270.0', '120.0', '650.0', '180.0', '1130.0',
       '450.0', '1640.0', '1460.0', '1020.0', '1030.0', '750.0', '640.0',
       '1070.0', '490.0', '1310.0', '630.0', '2000.0', '390.0', '430.0',
       '850.0', '210.0', '1430.0', '1950.0', '440.0', '220.0', '1160.0',
       '860.0', '580.0', '2060.0', '1820.0', '1180.0', '200.0', '1150.0',
       '1200.0', '680.0', '530.0', '1450.0', '1170.0', '1080.0', '960.0',
       '280.0', '870.0', '1100.0', '460.0', '1400.0', '660.0', '1220.0',
       '900.0', '420.0', '1580.0', '1380.0', '475.0', '690.

In [10]:
# sqft_basement to float
df['sqft_basement'] = df['sqft_basement'].astype('float')

In [11]:
# check if the transformation worked
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21143 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21143 non-null  int64  
 1   date           21143 non-null  object 
 2   price          21143 non-null  float64
 3   bedrooms       21143 non-null  int64  
 4   bathrooms      21143 non-null  float64
 5   sqft_living    21143 non-null  int64  
 6   sqft_lot       21143 non-null  int64  
 7   floors         21143 non-null  float64
 8   waterfront     18804 non-null  object 
 9   view           21082 non-null  object 
 10  condition      21143 non-null  object 
 11  grade          21143 non-null  object 
 12  sqft_above     21143 non-null  int64  
 13  sqft_basement  21143 non-null  float64
 14  yr_built       21143 non-null  int64  
 15  yr_renovated   17389 non-null  float64
 16  zipcode        21143 non-null  int64  
 17  lat            21143 non-null  float64
 18  long  

#### Checking and removing null values

Looking at the info on the data, we can see that the waterfront and the year renovate columns have missing values.

In [12]:
# Check for missing values
df.isna().sum()

id                  0
date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2339
view               61
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3754
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

In [14]:
# Create a function for calculating the percentage of missing values
def percentage_missing_values(data):
    percent_missing = data.isnull().sum() * 100 / len(data)
    missing_value_df = pd.DataFrame({'column_name': data.columns,
                                 'percent_missing': percent_missing,
                                 'number_missing': data.isnull().sum()})
    return missing_value_df

In [15]:
# Pass our data thorugh the function
percentage_missing_values(df)

column_name  percent_missing  number_missing
id                        id         0.000000               0
date                    date         0.000000               0
price                  price         0.000000               0
bedrooms            bedrooms         0.000000               0
bathrooms          bathrooms         0.000000               0
sqft_living      sqft_living         0.000000               0
sqft_lot            sqft_lot         0.000000               0
floors                floors         0.000000               0
waterfront        waterfront        11.062763            2339
view                    view         0.288512              61
condition          condition         0.000000               0
grade                  grade         0.000000               0
sqft_above        sqft_above         0.000000               0
sqft_basement  sqft_basement         0.000000               0
yr_built            yr_built         0.000000               0
yr_renovated    yr_renovated        17.755285            3754
zipcode              zipcode         0.000000               0
lat                      lat         0.000000               0
long                    long         0.000000               0
sqft_living15  sqft_living15         0.000000               0
sqft_lot15        sqft_lot15         0.000000               0

As expected, the waterfront (11.06%), yr_renovated (17.76%) and view (0.29%) columns have missing values. Since the data set has 21000 entries, dropping the rows with the missing values will lead to the loss of a chunk of data. Since we are not going to drop the roes or the columns, we are going to fill the rows with a common value for the said columns. We begin by investigating the statistics for the said columns so that we choose which value to use when filling the null values.

In [21]:
# Understanding the view column
df.view.head(30)

0          NONE
1          NONE
2          NONE
3          NONE
4          NONE
5          NONE
7           NaN
8          NONE
9          NONE
10         NONE
11         NONE
12         NONE
13         NONE
14         NONE
15         GOOD
16         NONE
17         NONE
19         NONE
20         NONE
21    EXCELLENT
22         NONE
23         NONE
24         NONE
25         NONE
26         NONE
27         NONE
28         NONE
29         NONE
30         NONE
31         NONE
Name: view, dtype: object

In [16]:
# Checking for the view column
df['view'].describe()

count     21082
unique        5
top        NONE
freq      19018
Name: view, dtype: object

In [23]:
df['view'].value_counts()

NONE         19018
AVERAGE        930
GOOD           496
FAIR           327
EXCELLENT      311
Name: view, dtype: int64

Upon further investigation of the view column, I found out that it contains over 19000 none values used as placeholder values. Due to this, I will drop the column because it will intefere with analysis and developing a model. 

In [25]:
# Drop the view column
df = df.drop('view', axis=1)

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21143 entries, 0 to 21596
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21143 non-null  int64  
 1   date           21143 non-null  object 
 2   price          21143 non-null  float64
 3   bedrooms       21143 non-null  int64  
 4   bathrooms      21143 non-null  float64
 5   sqft_living    21143 non-null  int64  
 6   sqft_lot       21143 non-null  int64  
 7   floors         21143 non-null  float64
 8   waterfront     18804 non-null  object 
 9   condition      21143 non-null  object 
 10  grade          21143 non-null  object 
 11  sqft_above     21143 non-null  int64  
 12  sqft_basement  21143 non-null  float64
 13  yr_built       21143 non-null  int64  
 14  yr_renovated   17389 non-null  float64
 15  zipcode        21143 non-null  int64  
 16  lat            21143 non-null  float64
 17  long           21143 non-null  float64
 18  sqft_l

In [30]:
# Exploring the waterfront column
print(df.waterfront.head())
print(df.waterfront.value_counts())

0    NaN
1     NO
2     NO
3     NO
4     NO
Name: waterfront, dtype: object
NO     18662
YES      142
Name: waterfront, dtype: int64


We can see that more than 18000 of our values have the value 'NO' for the waterfront which means that the house does not have waterfronts. Therefore I am going to assume that the rows with missing values are for house with no waterfront, and fill the missing values with 'NO'. 

In [33]:
df.isna().sum()

id                   0
date                 0
price                0
bedrooms             0
bathrooms            0
sqft_living          0
sqft_lot             0
floors               0
waterfront       21143
condition            0
grade                0
sqft_above           0
sqft_basement        0
yr_built             0
yr_renovated      3754
zipcode              0
lat                  0
long                 0
sqft_living15        0
sqft_lot15           0
dtype: int64

After trying to fill the column, I ended up with more missing values than before. I might have made a mistake with that so I am just going to drop the column because it will not be of help. 

In [34]:
# Drop the waterfront column
df = df.drop('waterfront', axis = 1)

In [35]:
# check the data frame preview
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21143 entries, 0 to 21596
Data columns (total 19 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21143 non-null  int64  
 1   date           21143 non-null  object 
 2   price          21143 non-null  float64
 3   bedrooms       21143 non-null  int64  
 4   bathrooms      21143 non-null  float64
 5   sqft_living    21143 non-null  int64  
 6   sqft_lot       21143 non-null  int64  
 7   floors         21143 non-null  float64
 8   condition      21143 non-null  object 
 9   grade          21143 non-null  object 
 10  sqft_above     21143 non-null  int64  
 11  sqft_basement  21143 non-null  float64
 12  yr_built       21143 non-null  int64  
 13  yr_renovated   17389 non-null  float64
 14  zipcode        21143 non-null  int64  
 15  lat            21143 non-null  float64
 16  long           21143 non-null  float64
 17  sqft_living15  21143 non-null  int64  
 18  sqft_l

Moving on to the yr_renovated column, we can get the summary statistics to understanf the column then use a value to fill the missing values

In [36]:
# Checking the summary statistics for the yr_renovated
df.yr_renovated.describe()

count    17389.000000
mean        82.991949
std        398.481819
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max       2015.000000
Name: yr_renovated, dtype: float64

In [37]:
# Checking the values in the column
df.yr_renovated.value_counts()

0.0       16666
2014.0       69
2003.0       31
2013.0       31
2007.0       30
          ...  
1953.0        1
1944.0        1
1934.0        1
1971.0        1
1959.0        1
Name: yr_renovated, Length: 69, dtype: int64

It seems like 16666 people put 0 for the year they got their house renovated. I assume that the 0 meant that they nver got their houses renovated. Since this feels like a safe assumption for our data, I am going to fill the missing values with the median which is 0. 

In [38]:
# Fill the missing values in yr_renovated column with the median
yr_renovated_median = df.yr_renovated.median()
df.yr_renovated = df.yr_renovated.fillna(value = yr_renovated_median)
df.head(10)

id        date      price  bedrooms  bathrooms  sqft_living  \
0   7129300520  10/13/2014   221900.0         3       1.00         1180   
1   6414100192   12/9/2014   538000.0         3       2.25         2570   
2   5631500400   2/25/2015   180000.0         2       1.00          770   
3   2487200875   12/9/2014   604000.0         4       3.00         1960   
4   1954400510   2/18/2015   510000.0         3       2.00         1680   
5   7237550310   5/12/2014  1230000.0         4       4.50         5420   
7   2008000270   1/15/2015   291850.0         3       1.50         1060   
8   2414600126   4/15/2015   229500.0         3       1.00         1780   
9   3793500160   3/12/2015   323000.0         3       2.50         1890   
10  1736800520    4/3/2015   662500.0         3       2.50         3560   

    sqft_lot  floors  condition          grade  sqft_above  sqft_basement  \
0       5650     1.0    Average      7 Average        1180            0.0   
1       7242     2.0    Average      7 Average        2170          400.0   
2      10000     1.0    Average  6 Low Average         770            0.0   
3       5000     1.0  Very Good      7 Average        1050          910.0   
4       8080     1.0    Average         8 Good        1680            0.0   
5     101930     1.0    Average   11 Excellent        3890         1530.0   
7       9711     1.0    Average      7 Average        1060            0.0   
8       7470     1.0    Average      7 Average        1050          730.0   
9       6560     2.0    Average      7 Average        1890            0.0   
10      9796     1.0    Average         8 Good        1860         1700.0   

    yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0       1955           0.0    98178  47.5112 -122.257           1340   
1       1951        1991.0    98125  47.7210 -122.319           1690   
2       1933           0.0    98028  47.7379 -122.233           2720   
3       1965           0.0    98136  47.5208 -122.393           1360   
4       1987           0.0    98074  47.6168 -122.045           1800   
5       2001           0.0    98053  47.6561 -122.005           4760   
7       1963           0.0    98198  47.4095 -122.315           1650   
8       1960           0.0    98146  47.5123 -122.337           1780   
9       2003           0.0    98038  47.3684 -122.031           2390   
10      1965           0.0    98007  47.6007 -122.145           2210   

    sqft_lot15  
0         5650  
1         7639  
2         8062  
3         5000  
4         7503  
5       101930  
7         9711  
8         8113  
9         7570  
10        8925